Most of the rest of the NU6.3 owl (#1319)
zcash/zips: zips/zip-0326.md

* We also explain how scanning for *Orchard-pool* funds should be handled across the upgrade.
* We describe how wallets should randomize note ciphertexts when spending under the
  *Orchard pool*'s post-NU6.3 same-address restriction, in order to maintain privacy against
  quantum and discrete-log-breaking adversaries.

# Motivation

For interoperability with existing addresses, *Ironwood-pool* notes are allowed to use
$\mathsf{use\_qsk} = \textsf{false}$ keys as well as $\mathsf{use\_qsk} = \textsf{true}$ keys.

In ZIP 2005, the incoming viewing key (and the diversifier key and outgoing viewing key) are
derived through $\mathsf{rivk}$, which depends on $\mathsf{use\_qsk}$; consequently the addresses
an account exposes, and the $\mathsf{ivk}$ needed to trial-decrypt notes sent to them, *also depend
on* $\mathsf{use\_qsk}$. A wallet that did not know which value applied — for example after
restoring from seed — would therefore have to trial-decrypt each *Ironwood-pool* note ciphertext
against *four* incoming viewing keys: the external and the internal, each for
$\mathsf{use\_qsk} \in \{\textsf{false}, \textsf{true}\}$.

> A wallet that created its own seed always knows which $\mathsf{use\_qsk}$ value is used, but

Reviewer: ```suggestion
A wallet that performed its own key derivation always knows which $\mathsf{use\_qsk}$ value is used, but
```

scanning cost close to its pre-NU6.3 level in the common cases while remaining correct in the
restore-from-seed case.

The recommendations here assume that pools and activations refer to Mainnet, and that a secure
wallet either does not support Testnet, or strictly segregates key material between Testnet and
Mainnet.

# Specification

## Scanning for Ironwood-pool and Orchard-pool notes

A receiver, and the corresponding incoming viewing key, is scoped to the Orchard *protocol*, not
to a pool: the same $\mathsf{ivk}$ is used to trial-decrypt both *Orchard-pool* and
*Ironwood-pool* note ciphertexts.

We assume that secure wallets do not expose *internal* receivers or viewing keys to any
party that is not subject to the same restrictions as the wallet itself. A wallet MUST NOT use
the fact that it knows the internal addresses of all of its accounts to send funds from one
account to an *internal* address of another.

## Wallet key-generation restrictions

The following restrictions make the scanning optimizations below sound. They concern only how a
wallet generates and uses keys; they are not consensus rules.

> * For each account, a wallet MUST generate keys with the same value of $\mathsf{use\_qsk}$ —

Reviewer: ```suggestion
* For each ZIP 32 account, a wallet MUST generate keys with the same value of $\mathsf{use\_qsk}$ —
```

  either all $\mathsf{use\_qsk} = \textsf{true}$ or all $\mathsf{use\_qsk} = \textsf{false}$,
  never a mixture.
* A wallet MUST NOT send funds to an external receiver in the *Orchard pool* after NU6.3

>   activation.

Reviewer: ```suggestion
* A wallet MUST NOT send funds to any external receiver (including its own) in the 
  Orchard pool* after NU6.3 activation.
```

* Any wallet that generates $\mathsf{use\_qsk} = \textsf{true}$ keys:
    * MUST NOT send *Orchard-pool* funds to them at any point; and
    * MUST NOT expose the corresponding Orchard-protocol receivers or viewing keys before NU6.3

>       activation.

Reviewer: ```suggestion
      activation, to avoid contamination of the Orchard pool.
```

Author: I would avoid the word "contamination" and say something more precise.

> would need the spending key, because the *Orchard pool*'s same-address restriction
> [^draft-zodl-valargroup-action-circuit-update] requires the spend side of an action to be at
> the same address as its output side, and control over the spending key is something that
> wallets can enforce. This property cannot be enforced by consensus — any party that knows a
> $\mathsf{use\_qsk} = \textsf{true}$ spending key can send funds to it regardless of the
> same-address restriction, since the spend side of the action can be a dummy — but the wallet
> restrictions above are sufficient.

</details>

### Scan ranges

Under the restrictions above, a wallet can reduce its scanning load and complexity:

* *Orchard-pool* note ciphertexts SHOULD NOT be scanned with $\mathsf{use\_qsk} = \textsf{true}$
  incoming viewing keys.

* For an account with birthday height after NU6.3 activation, the account will not have received
  *Orchard-pool* funds, so *Orchard-pool* note ciphertexts SHOULD NOT be scanned at all, with
  either the internal or the external $\mathsf{ivk}$.

* For an account with birthday height before NU6.3 activation, *Orchard-pool* note ciphertexts

>   SHOULD NOT be scanned with the *external* $\mathsf{use\_qsk} = \textsf{false}$ $\mathsf{ivk}$

Reviewer: Note that this doesn't work for wallets that send change back to the originating address (i.e. those that never adopted internal IVKs)

  after NU6.3 activation. (A transaction containing such an output would have to have been created
  by a non-conformant wallet that sent funds to an external *Orchard-pool* receiver after NU6.3.)

* If an account's *Orchard-pool* balance is known to be zero at a block at or after NU6.3
  activation, then *Orchard-pool* note ciphertexts SHOULD NOT be scanned in any block that
  descends from that block in the block tree. (After NU6.3, no new value may enter the *Orchard
  pool* [^zip-0258], so a zero balance there cannot subsequently become non-zero.)

Note that, for an account with birthday height before NU6.3, it might seem as though scanning
could be further optimized by not scanning the *Ironwood pool* with $\mathsf{use\_qsk} =
\textsf{true}$ incoming viewing keys. However, because a conservatively early birthday height has
traditionally been safe, such an optimization would *not* be safe — a $\mathsf{use\_qsk} =
\textsf{true}$ account whose recorded birthday height was set too early would have its

> $\mathsf{use\_qsk} = \textsf{true}$ *Ironwood-pool* notes skipped — and so it is not specified.

Reviewer: A scenario where this incorrect optimization would bite is that if a user restored a wallet and mistakenly set the birthday height too early, they could fail to find their funds.

  was consensus-valid on some chain.)

### First-funds scanning in the Ironwood pool

The rules so far do not, on their own, save a wallet that does not yet know its $\mathsf{use\_qsk}$
value from scanning with up to four incoming viewing keys in the *Ironwood pool* until it first receives
funds there. We describe an optimization for the *Ironwood pool* below. (It could be applied
separately to the *Orchard pool*, but that is less helpful, because the other rules almost subsume
the benefit.)

While the wallet knows that no *Ironwood-pool* funds have been received by the account before the
block being scanned, it only needs to use the *external* incoming viewing keys, with one exception: if
the transaction's `valueBalanceIronwood` is negative (value is entering the *Ironwood pool*), then
the transaction has an input to the pool via the value balance, which might go to an
*Ironwood-pool* note at an *internal* receiver of this wallet. (No other party can address an
internal receiver, so before the account has received *Ironwood-pool* funds, an internal note for
it can only arise from the account's own transaction, which must be funded either by spending

> existing *Ironwood-pool* funds — of which there are none yet — or by such an inflow.)

Reviewer 2: We should be doing this in scanning logic anyway! We were running with this in github.com/valargroup/librustzcash , until Ironwood. (There was one bug remaining, that we decided its easier to refactor post-Ironwood. So we removed that sync optimization for now)

Reviewer 2: So we are still doubling load for `use_qsk` in scanning, relative to what always should be the case. Its just that we have a 2x on the table for improvements within the pool for everyone

Reviewer 3: agreed, need to tighten up the different rules / preconditions being made here (that wallets can rely on without missing funds).

restrictions on generating and using $\mathsf{use\_qsk} = \textsf{true}$ keys take effect
immediately; the scanning rules apply from NU6.3 activation.

# References

[^BCP14]: [Information on BCP 14 — "RFC 2119: Key words for use in RFCs to Indicate Requirement Levels" and "RFC 8174: Ambiguity of Uppercase vs Lowercase in RFC 2119 Key Words"](https://www.rfc-editor.org/info/bcp14)

[^protocol]: [Zcash Protocol Specification, Version 2025.6.3 [NU6.1] or later](protocol/protocol.pdf)

[^protocol-networks]: [Zcash Protocol Specification, Version 2025.6.3 [NU6.1]. Section 3.12: Mainnet and Testnet](protocol/protocol.pdf#networks)

[^protocol-addressesandkeys]: [Zcash Protocol Specification, Version 2025.6.3 [NU6.1] or later. Section 3.1: Payment Addresses and Keys](protocol/protocol.pdf#addressesandkeys)

[^zip-0032]: [ZIP 32: Shielded Hierarchical Deterministic Wallets](zip-0032.rst)

[^zip-0200]: [ZIP 200: Network Upgrade Mechanism](zip-0200.rst)

[^zip-0229]: [ZIP 229: Version 6 Transaction Format](zip-0229.md)

[^zip-0258]: [ZIP 258: Deployment of the NU6.3 Network Upgrade](zip-0258.md)

[^zip-0307]: [ZIP 307: Light Client Protocol for Payment Detection](zip-0307.rst)

> [^zip-0317]: [ZIP 317: Proportional Transfer Fee Mechanism](zip-0317.rst)

Reviewer 3: orphan ref (https://github.com/zcash/zips/pull/1301#discussion_r3449297994).

Author: I will fix all orphan refs.

[^zip-2005]: [ZIP 2005: Ironwood Quantum Recoverability](zip-2005.md)

[^zip-2005-usagewithfrost]: [ZIP 2005: Ironwood Quantum Recoverability, section "Usage with FROST"](zip-2005.md#usagewithfrost)

[^pczt]: [zcash/zips issue #693: Standardize a protocol for creating shielded transactions offline](https://github.com/zcash/zips/issues/693)

> [^draft-schell-ironwood-migration]: [Orchard to Ironwood Migration (draft)](draft-schell-ironwood-migration.md)

Reviewer 3: what's the distinction between `draft-schell-ironwood-migration.md` and `draft-zodl-valargroup-ironwood-migration.md` which this seemingly uses interchangeably?

Author: This will be changed to ZIP 318 anyway.
